Lib/ntpath.py

```diff
@@ -640,23 +638,68 @@ def commonpath(paths):
         genericpath._check_arg_types('commonpath', *paths)
         raise
 
+MAX_PATH = 260
 
-# determine if two files are in fact the same file
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        is_str = isinstance(filename, str)
+        if not is_str:
+            filename = os.fsdecode(filename)
+        extended_path_prefix = '\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_path_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved if is_str else os.fsencode(resolved)
+
+    def _extended_to_normal(path):
+        letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        if (len(path) > 7 and
```

Comment:
This is one character too many. The path length should be greater than 6 for a logical drive and 7 for a UNC drive. For example, "\\\\?\\C:\\" (7) and "\\\\?\\UNC\\" (8). For the latter, the result will be "\\\\", which is fine because Windows itself will try to open "\\\\" as "\\??\\UNC\\".